services/ui_backend_service/data/db/tables/run.py

```diff
@@ -22,34 +23,83 @@ class AsyncRunTablePostgres(AsyncPostgresTable):
-                artifacts.task_id, artifacts.attempt_id, artifacts.ts_epoch,
-                attempt_ok.value::boolean as attempt_ok
-            FROM {artifact_table} as artifacts
-            LEFT JOIN {metadata_table} as attempt_ok ON (
-                artifacts.flow_id = attempt_ok.flow_id AND
-                artifacts.run_number = attempt_ok.run_number AND
-                artifacts.task_id = attempt_ok.task_id AND
-                artifacts.step_name = attempt_ok.step_name AND
-                attempt_ok.field_name = 'attempt_ok' AND
-                attempt_ok.tags ? ('attempt_id:' || artifacts.attempt_id)
-            )
-            WHERE artifacts.name = '_task_ok' AND artifacts.step_name = 'end'
-        ) AS artifacts ON (
-            {table_name}.flow_id = artifacts.flow_id AND
-            {table_name}.run_number = artifacts.run_number
-        )
+        LEFT JOIN LATERAL (
+            SELECT ts_epoch, value::boolean
+            FROM {metadata_table} as attempt_ok
+            WHERE
+                {table_name}.flow_id = attempt_ok.flow_id AND
+                {table_name}.run_number = attempt_ok.run_number AND
+                attempt_ok.step_name = 'end' AND
+                attempt_ok.field_name = 'attempt_ok'
+            ORDER BY ts_epoch DESC
+            LIMIT 1
+        ) as end_attempt_ok ON true
         """.format(
             table_name=table_name,
+            metadata_table=metadata_table
+        ),
+        """
+        LEFT JOIN LATERAL (
+            SELECT ts_epoch
+            FROM {metadata_table} as attempt
+            WHERE
+                {table_name}.flow_id = attempt.flow_id AND
+                {table_name}.run_number = attempt.run_number AND
+                attempt.step_name = 'end' AND
+                attempt.field_name = 'attempt' AND
+                end_attempt_ok.value IS FALSE
+            ORDER BY ts_epoch DESC
+            LIMIT 1
+        ) as end_attempt ON true
+        """.format(
+            table_name=table_name,
+            metadata_table=metadata_table
+        ),
+        """
+        LEFT JOIN LATERAL (
+            SELECT 1
+            FROM
+            {task_table} as task
+            LEFT JOIN LATERAL (
+                SELECT value::boolean as is_ok, ts_epoch
+                FROM {metadata_table} as attempt_ok
+                WHERE
+                    task.flow_id=attempt_ok.flow_id AND
+                    task.run_number=attempt_ok.run_number AND
+                    task.step_name=attempt_ok.step_name AND
+                    task.task_id=attempt_ok.task_id AND
+                    attempt_ok.field_name='attempt_ok'
+                LIMIT 1
+            ) AS attempt_ok ON true
+            WHERE
+                {table_name}.flow_id = task.flow_id
+                AND {table_name}.run_number = task.run_number
+                AND @(extract(epoch from now())-{table_name}.last_heartbeat_ts)>{heartbeat_threshold}
+                AND end_attempt_ok IS NULL
```

Comment:
I *think* you may be able to filter on `end_attempt IS NULL OR end_attempt_ok IS NULL` which should eliminate more checks (ie: as soon as `end` starts, it won't check this as opposed to `wait until we see an `attempt_ok` for `end`. Not sure it makes a huge difference but maybe?

Reply:
added the end_attempt IS NULL to the filter as well, didn't seem to have a significant impact though, but makes sense to take it along for the ride.

Reply:
I would have added as an OR but ya, it probably isn't hugely significant.